src/main/scala/fi/vm/sade/hakurekisteri/ovara/OvaraService.scala

```diff
@@ -0,0 +1,225 @@
+    try {
+      val count = saveInSiirtotiedostoPaged(params, pageFunction)
+      Right(count)
+    } catch {
+      case t: Throwable =>
+        logger.error(
+          s"(${params.executionId}) Virhe muodostettaessa siirtotiedostoa parametreilla $params:",
+          t
+        )
+        Left(t)
+    }
+  }
+
+  case class SiirtotiedostoEnsikertalainen(
+    hakuOid: String,
+    henkiloOid: String,
+    isEnsikertalainen: Boolean,
+    menettamisenPeruste: Option[MenettamisenPeruste]
+  )
+
+  //Haetaan hakujen oidit ja synkataan ensikertalaiset näille
+  def triggerEnsikertalaiset(vainAktiiviset: Boolean) = {
+    implicit val to: Timeout = Timeout(5.minutes)
+    val allHaut: Future[AllHaut] =
+      (hakuActor ? HakuRequest)
+        .mapTo[AllHaut] //fixme, tää on ongelma koska palautuu 0 hakua jos kaikkia hakuja ei oo vielä haettu actorin päässä.
+    val hakuResult = Await.result(allHaut, 5.minutes)
+    val kiinnostavat =
+      hakuResult.haut.filter(haku => (!vainAktiiviset || haku.isActive) && haku.kkHaku).map(_.oid)
+    logger.info(
+      s"Löydettiin ${kiinnostavat.size} kiinnostavaa hakua yhteensä ${hakuResult.haut.size} hausta. Vain aktiiviset: $vainAktiiviset"
+    )
+    formEnsikertalainenSiirtotiedostoForHakus(kiinnostavat)
+  }
+
+  //Ensivaiheessa ajetaan tämä kaikille kk-hauille kerran, myöhemmin riittää synkata kerran päivässä aktiivisten kk-hakujen tiedot
+  def formEnsikertalainenSiirtotiedostoForHakus(hakuOids: Seq[String]) = {
+    val executionId = UUID.randomUUID().toString
+    var fileNumber = 1
+    def formSiirtotiedostoForHaku(hakuOid: String) = {
+      implicit val to: Timeout = Timeout(30.minutes)
+
+      val ensikertalaiset: Future[Seq[Ensikertalainen]] =
+        (ensikertalainenActor ? HaunEnsikertalaisetQuery(hakuOid)).mapTo[Seq[Ensikertalainen]]
+      ensikertalaiset.map((rawEnsikertalaiset: Seq[Ensikertalainen]) => {
+        logger.info(
+          s"($executionId) Saatiin ${rawEnsikertalaiset.size} ensikertalaisuustietoa haulle $hakuOid. Tallennetaan siirtotiedosto."
+        )
+        val ensikertalaiset = rawEnsikertalaiset.map(e =>
+          SiirtotiedostoEnsikertalainen(
+            hakuOid,
+            e.henkiloOid,
+            e.menettamisenPeruste.isEmpty,
+            e.menettamisenPeruste
+          )
+        )
+        s3Client
+          .saveSiirtotiedosto[SiirtotiedostoEnsikertalainen](
+            "ensikertalainen",
+            ensikertalaiset,
+            executionId,
+            fileNumber
+          )
+      })
+    }
+
+    val infoStr =
+      if (hakuOids.size <= 5) s"hauille ${hakuOids.toString}" else s"${hakuOids.size} haulle."
+    logger.info(s"($executionId) Muodostetaan siirtotiedostot $infoStr")
+    val resultsByHaku = hakuOids.map(hakuOid => {
+      try {
+        val result = formSiirtotiedostoForHaku(hakuOid)
+        //Todo, muu toteutus tälle? Mikä on riittävä timeout, mitä jos jäädään jumiin? Käsiteltäviä hakuja voi olla paljon,
```

Comment:
Pitäiskö näille todoille tehdä jotain? Näitähän ei saa selville kuin testaamalle

Reply:
OK-567 selkiyttää tämänkin, pitää toimia ajastetusti järkevässä ajassa ja järkevillä resursseilla